### ISSUE TITLE
(Proposal) Change how users create/open projects and scenes

### ISSUE THREAD

**Issue by morevnaproject** [opentoonz/opentoonz#3259]
I am proposing to change the way how user opens/creates projects/scenes in OpenToonz.

What I feel wrong with current concept:
- OpenToonz defines a limited list of locations for projects. It is not possible to create a project in arbitrary location (in fact you can create a project in any of defined locations and then move it, but this is not convenient/intuitive).
- OpenToonz displays dialog at startup, which looks complicated for new users - to open a scene, it is required to select project first, then scene (two steps).

That's all different from what we used to in other applications.

For example in Blender - user creates a file and it already have scene inside. So user do not care about scenes unless he wants to create a new one. When user opens file Blender automatically activates scene, which was used last time in this file. So we can think of **"Blender file" = "OpenToonz Project"**.

I am proposing to rework the UX of OpenToonz to be like this:
1. The user starts OT, a dialog appears, with only two buttons: **create** or **open** a project + a list of last opened projects/scenes (see [comment](https://github.com/opentoonz/opentoonz/issues/3259#issuecomment-629638535)).
1. The user selects "create."
1. The application suggest to select the name of new project and the folder where to create the project (any folder, and not some kind of choice from 2-3 locations!)
1. The user sets the name and location. Clicks "Create."
1. The application creates the project **and** immediately creates one scene in it (with the default name) **and** immediately opens it.
1. So, user can start working immediately. If necessary - he creates new scenes (via "file" - "new Scene" and works with them.
1. The user finishes working, saves scene and quits application. The project remembers which scene was opened last.
1.  The user opens the application again, a dialog appears: **create** or **open** a project.
1. The user selects "open".
1. Application offers a dialogue in which you can go to any folder and choose which project to open from this folder 9you can see how this works in NetBeans, for example). Ni this dialog OT projects are displayed like "files" - user can select them and click open.
1. OT opens the project **and** immediately opens the scene that was last active.
1. The user is working.
1. And so on.

With this concept the workflow is project-centric and user do not thinks about scenes too much  unless this is needed.

**Comment by morevnaproject** [comment]
References from other applications:

For step 1, something similar in Krita:
![screenshot_001](https://user-images.githubusercontent.com/332868/80483425-43879680-8980-11ea-9e50-cf89a47eb869.png)

For 10 (opening project), here is a reference from PyCharm IDE (you can see some folders marked with special icon - they are projects):
![screenshot_002](https://user-images.githubusercontent.com/332868/80483713-bc86ee00-8980-11ea-9eec-11deed17e7a2.png)

Similar window for opening a project in NetBeans IDE:
![screenshot_003](https://user-images.githubusercontent.com/332868/80483855-f657f480-8980-11ea-922c-0f10d25ac9af.png)

Scenes in Blender:
![screenshot_004](https://user-images.githubusercontent.com/332868/80484488-fc9aa080-8981-11ea-9a09-ad58726e622b.png)

**Comment by age3rcm** [comment]
i still have trouble trying to renember the start dialog procedure
some times i think i got it right and sometimes i still screw something
a more intuitive start box is supported

i would say something to do with the "file" menu too, the options fills up a 1080P screen from top to bottom

**Comment by konero** [comment]
@age3rcm The menus were updated recently, see: #2545

**Comment by ave4** [comment]
> OT opens the project and immediately opens the scene that was last active.

And if I want to load another scene?

I am also against opentoonz creating the scenes myself. I think it’s better to add options for auto-creation of scenes according to the template in the program settings. For example scene_ $ n

**Comment by morevnaproject** [comment]
>> OT opens the project and immediately opens the scene that was last active.
>
> And if I want to load another scene?

You can just switch scene. 
It would be nice to add scene switch widget like in Blender (see screenshot from https://github.com/opentoonz/opentoonz/issues/3259#issuecomment-620561402).
Or, while this widget is not available you can use existing functional: select "File" -> "Load Scene". ^__^

**Comment by age3rcm** [comment]
> @age3rcm The menus were updated recently, see: #2545

i guess... i forgot about it, too...
maybe its because i am watching too much old tutorials, sorry

**Comment by turtletooth** [comment]
I'm not sure what I think about this.  This would help OpenToonz with the beginner audience who is moving up from Medibang or Sai, but the options on the startup page give fast access to experienced users.  The ability to get to previous work quickly or switch projects is important.  

What audience do we want to cater to?  All of the concepts on the current startup screen are things that people will need to understand if they are going to get into animation.

There's really only one entry that I would want to see removed - the units.  I wish that pixels was the default.  That can be a real confusing piece, especially for pure digital animators.

**Comment by morevnaproject** [comment]
Hi Jeremy! ^__^

>  The ability to get to previous work quickly or switch projects is important.

The proposed approach provides the same, some operation even faster.
Maybe I missed some common operation - please let me know about that.

>  All of the concepts on the current startup screen are things that people will need to understand if they are going to get into animation.

Arguable. People successfully use other animation programs without digging into complicated internals. Those internals surely exists in many apps (i.e. in Krita: their files are organized very similarly to OpenToonz projects, they just have them zipped and renamed into ".kra" extension - just try to rename Krita file into .zip ~_^).

I am making a free educational course about OpenToonz (draft, in Russian - https://morevna-school.ru/programma/opentoonz/) and if the approach above will be implemented, then I know how to slowly "unwrap" the concept of projects to end user instead of slapping all technical details into newbie's face from the start. ^__^

**Comment by artisteacher** [comment]
I feel a little mixed about this myself. I do think that new users do get very confused by the way projects work. I wonder if it needs to be quite so limiting. 

I would like the ability to easily create a project anywhere. The current system is definitely confusing to students.

I would also like a preference to set the default project location of your choice. This would be beneficial for studios and for teachers dealing with disorganized students 🙃

I strongly prefer having quick access to recently open scenes in the startup window. I often have several scenes in progress at the same time. It would slow me down if I needed to get into the last open scene in order to switch to the one I actually intended to open.

**Comment by morevnaproject** [comment]
> I wonder if it needs to be quite so limiting.

There are no limiting of functionality here. We just make it less techy and more user-friendly. 

> I would also like a preference to set the default project location of your choice. This would be beneficial for studios and for teachers dealing with disorganized students

Yes, that's good idea - can be an option in preferences.

> I strongly prefer having quick access to recently open scenes in the startup window. I often have several scenes in progress at the same time. It would slow me down if I needed to get into the last open scene in order to switch to the one I actually intended to open.

For switching scenes I suggest to implement a drop-down menu, like in Blender:
> Scenes in Blender:
> ![screenshot_004](https://user-images.githubusercontent.com/332868/80484488-fc9aa080-8981-11ea-9a09-ad58726e622b.png)

In this case the switching of scene can be done in 2 clicks.

I understand, many of us are working with several scenes and several projects.
Can you please describe what is our typically switch scenes/projects during your work?

**Comment by artisteacher** [comment]
> In this case the switching of scene can be done in 2 clicks.

Not if you are just opening OT. 1 click to enter the project. 1 click to open the scene dropdown. 1 click to select the scene. 3 clicks to choose a recent scene instead of the 1 click we have now.

I wouldn't mind being able to change scenes using the Blender setup once a project/scene is already open. I really wouldn't want that to replace scene selection in the startup window.

**Comment by gab3d** [comment]
@morevnaproject 
Please note the Blender "scene loading" example you show is not really a scene loading workflow, but a "scene display" one (as .blend files can contain several scenes embedded, and those embedded scenes inside the currently opened .blend file, are the only ones shown in the drop-down).
This is not the case in OT, as we can only have one scene open at any time.

**Comment by morevnaproject** [comment]
> I wouldn't mind being able to change scenes using the Blender setup once a project/scene is already open. I really wouldn't want that to replace scene selection in the startup window.

@artisteacher Thank you for explaining. Then, we can display a list of last scenes (with names of projects).

> References from other applications:
> 
> For step 1, something similar in Krita:
> ![screenshot_001](https://user-images.githubusercontent.com/332868/80483425-43879680-8980-11ea-9e50-cf89a47eb869.png)

Here, in "Recent Documents" section we can display a pairs of project-name + scene name:

* PROJECT1 / SCENE1
* PROJECT1 / SCENE3
* PROJECT2 / SCENE1
* ...

> Please note the Blender "scene loading" example you show is not really a scene loading workflow, but a "scene display" one (as .blend files can contain several scenes embedded, and those embedded scenes inside the currently opened .blend file, are the only ones shown in the drop-down; btw they're always opened as long as the .blend file is open, it's just that you can't see its content in the editors until you make one of them the current or active one).
> 
> This is not the case in OT, as we can only have one scene open at any time.

Yes, from technical point of view the opening scene in Blender and opening scene in Synfig - two different processes.
But from conceptual point of view they are the same:
* .blend file = OT project
* scene in blend file = scene in OT project

I would like to emphasize, that Blender file is conceptually the same as OT project (folder).

**Comment by morevnaproject** [comment]
> I wouldn't mind being able to change scenes using the Blender setup once a project/scene is already open. I really wouldn't want that to replace scene selection in the startup window.

@artisteacher Thank you for explaining. Then, we can display a list of last scenes (with names of projects).

> References from other applications:
> 
> For step 1, something similar in Krita:
> ![screenshot_001](https://user-images.githubusercontent.com/332868/80483425-43879680-8980-11ea-9e50-cf89a47eb869.png)

Here, in "Recent Documents" section we can display a pairs of project-name + scene name:

* PROJECT1 / SCENE1
* PROJECT1 / SCENE3
* PROJECT2 / SCENE1
* ...

> Please note the Blender "scene loading" example you show is not really a scene loading workflow, but a "scene display" one (as .blend files can contain several scenes embedded, and those embedded scenes inside the currently opened .blend file, are the only ones shown in the drop-down; btw they're always opened as long as the .blend file is open, it's just that you can't see its content in the editors until you make one of them the current or active one).
> 
> This is not the case in OT, as we can only have one scene open at any time.

@gab3d  Yes, from technical point of view the opening scene in Blender and opening scene in Synfig - two different processes.

But from conceptual point of view they are the same:
* .blend file = OT project
* scene in blend file = scene in OT project

I would like to emphasize, that Blender file is conceptually the same as OT project (folder).

**Comment by morevnaproject** [comment]
> I wouldn't mind being able to change scenes using the Blender setup once a project/scene is already open. I really wouldn't want that to replace scene selection in the startup window.

@artisteacher Thank you for explaining. Then, we can display a list of last scenes (with names of projects).

> References from other applications:
> 
> For step 1, something similar in Krita:
> ![screenshot_001](https://user-images.githubusercontent.com/332868/80483425-43879680-8980-11ea-9e50-cf89a47eb869.png)

Here, in "Recent Documents" section we can display a pairs of project-name + scene name:

* PROJECT1 / SCENE1
* PROJECT1 / SCENE3
* PROJECT2 / SCENE1
* ...

> Please note the Blender "scene loading" example you show is not really a scene loading workflow, but a "scene display" one (as .blend files can contain several scenes embedded, and those embedded scenes inside the currently opened .blend file, are the only ones shown in the drop-down; btw they're always opened as long as the .blend file is open, it's just that you can't see its content in the editors until you make one of them the current or active one).
> 
> This is not the case in OT, as we can only have one scene open at any time.

@gab3d  Yes, from technical point of view the opening scene in Blender and opening scene in OpenToonz - two different processes.

But from conceptual point of view they are the same:
* .blend file = OT project
* scene in blend file = scene in OT project

I would like to emphasize, that Blender file is conceptually the same as OT project (folder).

**Comment by morevnaproject** [comment]
@artisteacher I have updated statement number 1 in my proposal (top post in this thread).

**Comment by phillycheese007** [comment]
"It is not possible to create a project in arbitrary location"
what about preferences "custom project path" " "multiple paths can be separated by **"

or just create folders and then create your projects from there ..if you want them to pop out in the browser add them in the prefs

**Comment by ave4** [comment]
You can specify in the settings "custom project path" as the path / for example in Linux. Then you can choose absolutely any path in toonz.

**Comment by morevnaproject** [comment]
> You can specify in the settings "custom project path" as the path / for example in Linux

Good point! But you can't do that on Windows - you have to add each disk if you have several disks.

> what about preferences "custom project path" " "multiple paths can be separated by **"
> 
> or just create folders and then create your projects from there ..if you want them to pop out in the browser add them in the prefs

Yes, it IS possible, but after tweaking configuration.
My reason is "why have such limit, while other apps do not have it?" ^__^

**Comment by shun-iwasawa** [comment]
One point I would like to remind you is that the "project file" and the "assets for the projects" are not needed to be stored in the same location.

You can specify any location for the assets by specifying the path for aliases (like `+drawings`, `+scenes`, etc.) in the project settings.

In short, the project file (`projectname_otprj.xml`) has only two information.
- File paths to the asset folders. ( like `+drawings = C:\path\to\drawings` )
- Settings to be used for the default value when creating a new scene under the project.

And the project root is just for storing the project files, not assets.

I'm unfamiliar with Blender, but I think you can say that OpenToonz project is correspond to one title of the film (or title of TV series).
For instance, in our studio the project is created only once at the beginning of the production of each film. After that, during the production period they only create scenes and assets which belong to the project.

Of course the usage can vary according to users, but IMO creating a new project may not be frequently selected compared to creating or loading scene.

**Comment by morevnaproject** [comment]
@shun-iwasawa I understand that for big studios the project creation procedure is not frequent (especially comparing to amount of other work).

But I have hit completely different situation when teaching animation to teenagers.

First of all, they often want to create something quick-and-simple, following their inspiration. Some short animation, like one below.

![2019-12-31-NG](https://user-images.githubusercontent.com/332868/83602120-286dff00-a59c-11ea-81b8-857036de85eb.gif)

They just want to open application, create file, draw, save and close.

Here are common problems that I experienced in my practice:

1. Students do not know where their files located. They creating a project, but they often puzzled "why app doesn't asking where to save it?" The look of "New Project" dialog is uncommon for them and they usually can't "read" it properly.

![screenshot_001](https://user-images.githubusercontent.com/332868/83602482-d4174f00-a59c-11ea-8f58-41abf3de8cf3.png)

2. In studio we have pre-configured additional locations to store projects. But when student comes home, he often have troubles finding his files. 

3. Students often forget to choose correct project before he starts working. He just creates a new scene and it gets saved in the wrong project then. It is often tricky to separate his files and transfer to other project (student is not able to do that without teacher's help),

![screenshot_002](https://user-images.githubusercontent.com/332868/83604602-a2a08280-a5a0-11ea-9071-6c68e1b0aa62.png)

4. There are worst situations can take place - student launches OpenToonz and just closing Startup Popup (without choosing anything). In this case he works with project, which was opened last. When he hits Ctrl+S then he see a dialog suggesting to save a scene somewhere. He realizing that this is a wrong project and selecting location of his own project. But all drawings (level files) are remaining in previous project and thus a confusion happen when student wants to copy his work somewhere (i.e. when he wants to take his work home).

All those issues related with current complexity of creating projects.
Consider, that our students are at age 14-17 and they all dream to become animation artists, but due to their age they are not yet ready to dig into very technical details as current procedure of project creation. Even some of my teacher colleagues get confused sometimes.

> You can specify any location for the assets by specifying the path for aliases (like +drawings, +scenes, etc.) in the project settings.

I am aware about flexibility offered by OT's project concept and in fact I find it very impressive. I like the fact that I can separate assets of each scene by using "append $scenepath" option. I also like that I can set arbitrary locations for assets.

![screenshot_003](https://user-images.githubusercontent.com/332868/83605711-94ebfc80-a5a2-11ea-9e54-9f1c5086299f.png)

What I am proposing is to make the procedure of opening project is more explicit.

Currently in OT the most explicit operation is opening a **scene**:
- User launches OT -> opens scene -> scene opened (project is switched implicitly)

I am proposing to make project opening explicit:
- User launches OT -> chooses project to open -> OT switches to this project and opens its last used scene

**Comment by ave4** [comment]
The right decision here was to teach the steps to work with projects. Still, if they are engaged in animation, then they still have to work with large projects, for example, doing professional animation.

> User launches OT -> chooses project to open -> OT switches to this project and opens its last 

Or offers to select any scene from the project

**Comment by morevnaproject** [comment]
> The right decision here was to teach students how to work with projects.

What I want to say is that currently the way how user works with projects is counter-intuitive.
That prevents people from using OT for doing some simple-and-quick.

> Even some of my teacher colleagues get confused sometimes.

**Comment by shun-iwasawa** [comment]
> 1. Students do not know where their files located. They creating a project, but they often puzzled "why app doesn't asking where to save it?" The look of "New Project" dialog is uncommon for them and they usually can't "read" it properly.

I'm agree with you that the current UI of "New Project" dialog is quite confusing. Users may  wrongly detect the value in each field of path aliases as the folder name. But it is actually a relative path from the project file which they can be replaced by any location. It has room for improvement.


> 2. In studio we have pre-configured additional locations to store projects. But when student comes home, he often have troubles finding his files.
> 3. Students often forget to choose correct project before he starts working. He just creates a new scene and it gets saved in the wrong project then. It is often tricky to separate his files and transfer to other project (student is not able to do that without teacher's help)
> 4. There are worst situations can take place - ...

As you pointed, moving assets in one project to another location without breaking references is not easy. Because the project is basically managed in absolute paths.

Instead, it might be worth considering using the Scene Folder Alias.
It can be enabled by setting the preferences option `General -> Path Alias Priority` to `Scene Folder Alias ($scenefolder)` .

The scene folder alias provides users to non-project dependent way for managing assets. All assets (drawings, backgrounds, etc.) for a scene will be saved in one folder (= scene folder) and you will keep the references even if you move the folder to another location, because all references will be managed in relative paths (like `$scenefolder/A.tlv`) from the folder containing the current scene file (.tnz file).
I think this way would be more suitable for smaller projects, and for users sharing one computer with the same user name.

Please note that the `$scenefolder` alias is only available if the current scene is saved in somewhere. We may need some mode for beginners that the software won't start with the current scene left untitled in order to avoid the trouble.


Getting back to the original post...

> 3. The application suggest to select the name of new project and the folder where to create the project (any folder, and not some kind of choice from 2-3 locations!)

I think it is possible to modify the software enabling the project files to be saved in anywhere outside the project root so that it gains portability. The drawback would be that such projects won't be listed in the File Browser tree or in widgets for selecting projects, since OT has no way of tracking the locations of them.

**Comment by ave4** [comment]
It would be nice to force opentoonz to open tnz files with a simple mouse click on it in the OS file manager

**Comment by RodneyBaker** [comment]
@ave said:
It would be nice to force opentoonz to open tnz files with a simple mouse click on it in the OS file manager.

That is more of an operating system setting than anything else.
On Windows there are several ways we can set this up.
Perhaps the most basic approach is to Right Click on a .TNZ scene file and select "Open with...".
At this point choose the 'Other Application' option even if Opentoonz is listed as an option because we want to specifically set up additional options and make sure Opentoonz always opens .TNZ scene files (not just for that instance).  Go through the options and set Opentoonz as the program that works with .TNZ files and you should see .TNZ files take on the Opentoonz icon in the file menu.  Then you can double click on the file to open it directly in Opentoonz.

Note that single clicking to open a file is a different setting and I'm not sure I'd recommend that.

**Comment by morevnaproject** [comment]
> We may need some mode for beginners that the software won't start with the current scene left untitled in order to avoid the trouble.

This is a great idea, I have problems with unsaved scenes regularly and this can be a solution.
I have created a separate issue for that. ^__^

> It can be enabled by setting the preferences option General -> "Path Alias Priority" to "Scene Folder Alias ($scenefolder)".

Thank you, I wasn't aware about that feature!
Actually I do not like "projectless" approach, as I find project's concept of OpenToonz very powerful. I just want to make it simpler for end user.
But I need to play with $scenefolder feature and see how it works...

**Comment by artisteacher** [comment]
> The drawback would be that such projects won't be listed in the File Browser tree or in widgets for selecting projects, since OT has no way of tracking the locations of them.

What if we could "pin" projects in the browser sidebar or etc.? On macOS's Finder, we can drag any directory to the sidebar for quick access.

**Comment by RodneyBaker** [comment]
> What if we could "pin" projects in the browser sidebar or etc.?

Not quite the same but (amongst other places) scenes get listed for easy retrieval in the Startup popup window and the Recently used Scene listing.

Keep in mind that in general with Opentoonz when we are talking about Projects we are talking about Scenes... just specifically with the additional criteria of the directory structure related to the scene coming along for the ride.

And... (conjecture on)...
This would be a rather radical change but changing the term 'Project' to 'Production' might help to shift the paradigm of how users look at the current structure just enough to begin to remedy the underlying problems associated with users thinking in terms of other applications and their use of 'projects'.  This would further align the process so that a 'pre-pro-post' categorization of tasks could be considered.  Namely that the core production is done inside of the Opentoonz application with files inside the production directory while pre-production and post-production deals largely with the handoff from/to other applications.  (read that also as input>production>output as necessary).
A 'production' file (i.e. zipped up collection of the contents of an Opentoonz production with all of it's assets/resources could be a .pro, .prod or .production file.  That same file renamed with a different extension could signal which level of the production it was currently assigned to (.pre, .pro. and .post).  For that last one a .pos extension might be the norm until final where the trailing letter 't' is added to indicate the termination/completion of the production.

**Comment by gab3d** [comment]
Just as a reference here. This is the Maya New Project window, which is pretty much similar to OpenToonz one:

![imagen](https://user-images.githubusercontent.com/18104091/84548392-e3547600-acdb-11ea-8c42-bea338406e7a.png)

I don't see people flying away from Maya because of this... 

I mean, yeah! let's do OT project managing as easy and user friendly as possible, but please let's not lose the good features that an organized project structure allows and facilitates.
Let's welcome freelancers and hobbyists without losing more structured productions and studios.
🙏

**Comment by turtletooth** [comment]
FYI- here's how I'm doing this in Tahoma:
![proj](https://user-images.githubusercontent.com/4576381/84549604-a2516700-acc5-11ea-8fa2-0cbe912321e6.PNG)

![newProj2](https://user-images.githubusercontent.com/4576381/84549609-a4b3c100-acc5-11ea-8642-825c78eb4896.PNG)

Projects can be made anywhere.  I currently have customizing the inside folders turned off, but it's pretty easy to turn back on.  Having project flexibility is doable in OT.

**Comment by gab3d** [comment]
@turtletooth 
what is Tahoma?

**Comment by turtletooth** [comment]
It's my fork of OpenToonz.

**Comment by morevnaproject** [comment]
> FYI- here's how I'm doing this in Tahoma:
> ![proj](https://user-images.githubusercontent.com/4576381/84549604-a2516700-acc5-11ea-8fa2-0cbe912321e6.PNG)
> 
> ![newProj2](https://user-images.githubusercontent.com/4576381/84549609-a4b3c100-acc5-11ea-8642-825c78eb4896.PNG)
> 
> Projects can be made anywhere. I currently have customizing the inside folders turned off, but it's pretty easy to turn back on. Having project flexibility is doable in OT.

@turtletooth Thank you very much! This looks great!

**Comment by RodneyBaker** [comment]
As there has not been any specific movement in this proposal, closing as part of general housekeeping.
There have been quite a few excellent ideas suggested here and some have even been implemented.
Let's reconsider this proposal in light of review of current requirements in the 2022 development cycle.
